### PR TITLE
Added projectile sprite testing to the CheckYaml lint rules

### PR DESCRIPTION
**Pull request by Mailaender**
A followup of https://github.com/OpenRA/OpenRA/pull/7709 for projectile sprites. This can detect crashes like #8219 and #8020 beforehand.

### CODE REVIEW

**Comment by abcdefg30** [comment]
Needs a rebase.

**Comment by Mailaender** [comment]
Rebased.

**Comment by reaperrr** [comment]
Needs another rebase.

**Comment by reaperrr** [comment]
Doesn't appear to work properly. I added `Sequence: idle2` to TD `70mm` cannon, which correctly crashed the game when firing, but it wasn't caught by `make test`.
I then did the same for the `Rockets` weapon, that one was caught but only on 'Take Bialystok', where the weapon is customised.

**Comment by Mailaender** [comment]
Rebased. Not sure why the weapon rule map yaml merging does not work.

**Comment by GraionDilach** [comment]
Can it be related to Inherits doesn't work in weapons.yaml neither?

**Comment by abcdefg30** [comment]
Except the thing @reaperrr pointed out, this seems to work. :+1: / :white_check_mark:

**Comment by pchote** [comment]
Needs a rebase.

File: OpenRA.Mods.RA/Graphics/TeslaZapRenderable.cs [hunk 1]
```diff
@@ -35,13 +35,15 @@ struct TeslaZapRenderable : IRenderable, IFinalizedRenderable
...
+		readonly string dimSequence;
+		readonly string brightSequence;
 		readonly int brightZaps, dimZaps;
 
 		WPos cachedPos;
 		WVec cachedLength;
 		IEnumerable<IFinalizedRenderable> cache;
 
-		public TeslaZapRenderable(WPos pos, int zOffset, WVec length, string image, int brightZaps, int dimZaps, string palette)
+		public TeslaZapRenderable(WPos pos, int zOffset, WVec length, string image, int brightZaps, int dimZaps, string palette, string dim, string bright)
```
Comment (pchote) [inline OpenRA.Mods.RA/Graphics/TeslaZapRenderable.cs]: The order of the arguments here is inconsistent: it specifies brightZaps followed by dimZaps, but then dim followed by bright.

To improve readability please change this to `public TeslaZapRenderable(WPos pos, int zOffset, WVec length, string image, string brightZapsSequence, int brightZapsCount, string dimZapsSequence, int dimZapsCount, string palette)`

**Comment by Mailaender** [comment]
Rebased. I don't understand why https://github.com/OpenRA/OpenRA/pull/8732 was merged before or pulled at all. It seemed to only include my bullet cleanup commit.

**Comment by pchote** [comment]
Confirmed @reaperrr's discovery that the lint test doesn't work for Bullets.  That will need to be fixed before we can merge this.

**Comment by Mailaender** [comment]
Updated.

**Comment by Mailaender** [comment]
I can't reproduce any problems. My test cases were #8219 and #8020.

**Comment by reaperrr** [comment]
It works on maps that customize weapons.

It does not work for the weapon yamls.

**Comment by pchote** [comment]
My test case is the same as what @reaperrr mentioned above: Add `Sequence: bogus` to a projectile in the weapon yaml, and observe that the tests fail.  Launch the game, and observe that it crashes when they are spawned.  There is no merging going on.

**Comment by pchote** [comment]
After additional testing I can narrow this down even further: the lint will pick up the bogus sequences *only* if a map modifies the same weapon.  Regular maps with no overrides do not detect the problem.

**Comment by Mailaender** [comment]
Are you sure that Image is != null for your Sequence: bogus?

**Comment by Mailaender** [comment]
As already told: I don't think there is anything I can do about this limitation.

**Comment by reaperrr** [comment]
I suspect that weapon rules are simply not treated as part of general rules, but anything custom on maps is treated as map rules hence why it works there.

File: OpenRA.Mods.Common/Lint/CheckSequences.cs [hunk 2]
```diff
@@ -89,6 +90,48 @@ public void Run(Action<string> emitError, Action<string> emitWarning, Map map)
 						}
 					}
 				}
+
+				foreach (var weaponInfo in map.Rules.Weapons)
+				{
+					var projectileInfo = weaponInfo.Value.Projectile;
+					if (projectileInfo == null)
+						return;
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckSequences.cs]: Some fairly straightforward debugging pointed me here.  This cancels the lint pass as soon as you encounter a weapon without a projectile.

Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckSequences.cs]: The very first weapon definition in TD triggers this case, so the pass is not run at all.

Comment (Mailaender) [inline OpenRA.Mods.Common/Lint/CheckSequences.cs]: Good catch. This was stupid.

**Comment by reaperrr** [comment]
Works now :+1:

**Comment by Mailaender** [comment]
:raised_hands:

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/550856c1b1281206b40e42d36e5ec157579beb73...d41f55e18f5f73c34eb9cb7064cdc9844ce2a17f)

**Pull request merged**